core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -145,49 +143,43 @@ private FileIO io(String location) {
       return io;
     }
 
-    synchronized (ioInstances) {
-      // double check while holding the lock
-      io = ioInstances.get(impl);
-      if (io != null) {
-        return io;
-      }
+    return ioInstances.computeIfAbsent(
```

Comment:
We do this after checking `get(key) == null`. So the performance reported by https://bugs.openjdk.java.net/browse/JDK-8161372 should not be a problem.